README.md
# AwakeMUD Community Edition
A fork of the Awakened Worlds MUD codebase. Pull requests welcome!

Join us at mudtest.mooo.com port 4000 to participate in playtesting.

## Features
- Significant reductions in memory leaks
- Significantly reduced ticklag
- A slew of bugfixes to everything from combat code to sound propagation

## OS Support
Tested on:
- OSX 10.12
- Ubuntu 14.04.2 LTS
- Raspbian Jessie

## Installation
- Install MySQL 5.
- Clone this repository to your machine.
- Run `SQL/gensql.sh` (or do the steps manually if it doesn't support your OS).
- Copy the mysql_config.cpp file to src.
- Edit src/Makefile and select the OS that looks closest to yours.
- From the src directory, run `make clean && make`.
- Change to the root directory and run the game (ex: `gdb bin/awake`, or `lldb bin/awake` on OS X).


changelog.txt
Information about new features will be written into this changelog, as will information about major bugfixes. Minor bugfixes and related code-quality changes can be found in the commit log (git log).


2017/06/17:
- Efficiency overhaul to bitfield.cpp has significantly accelerated tick performance.
- More memory leak fixes. See commit log.

2017/06/16:
- More memory leak fixes. See commit log.
- Further updates to SQL script to allow compatibility with MySQL < 5.7.8.

2017/06/15:
- Updated SQL script to overwrite the old DB when run and to allow for better error recovery.
- More memory leak fixes. See commit log.

2017/06/14:
- More memory leak fixes. See commit log.

2017/06/13:
- Removed a plethora of memory leaks.
- Fixed house loading code and the distributed house file; the boot phase is now more reliable.
- Added first-pass fix for brand-new characters being glitched by an immediate copyover.

2017/06/12:
- Large overhaul to warnings and code quality. Details in commit log.

2017/06/11:
- Vehicles can now be PURGEd by staff.

2017/06/10:
- Forced behavior after copyovers is now much more reliable.

2017/06/09:
- Poofs being unexpectedly (null) will no longer occur.

2017/06/06:
- .gitignore now won't try to check in your vehicle files.
- Added in skeleton of spellset command for staff (still needs enhancement, see Issue #36)
- Fixed typos in poofin/poofout defaults.

2017/06/04:
- Updates to credits and various other textfiles.
- Fixes for vehicles not saving in certain circumstances after purchase/transfer/etc.

2017/06/03:
- Added copyover.dat to the .gitignore file.

2017/06/02:
- Fixed the majority of the deprecated-conversion-from-const-char-to-char warnings.
- Fixed the majority of -Wall warnings.
- Pulled in old spec-procs from 0.8.16B.

2017/06/01:
- Added note board to the Immortal HQ at 10000 for staff collaboration.

2017/05/30:
- Further combat rewrites.

2017/05/29:
- Initial combat rewrite to eliminate bugs.

2017/05/28:
- Added DB cleanup to shutdown. Also added potential recovery for DB-has-gone-away issue.

2017/05/26:
- Fix for blank skill issue for immortals.

2017/05/25:
- Moved old log() function to log_vfprintf(); added new log function. See Issue #2.
- Clarified LIT room tag to read as STREETLIGHTS.
- Fixed SQL sanitization issues.
- Added some staff-assistance files to doc/staffdoc (thanks Finster!)

2017/05/24:
- Built the Immortal HQ for centralized testing of code. Start room is 10000, or GOTO IMMHQ.
- Made the "Privileged operation." message more user-friendly.
- Brought in changes from the old snippets thread, including:
  - Stealing requires PK on both parties
  - Peepholes are built into apartment doors (LOOK <dir>)
- Added more bugfixes, including to ranged response code and the time-played function.

2017/05/23:
- NPCs will no longer keep closing the distance when their target is not there.
- Updated ranged response code for sanity, including messaging.
- Silencers and suppressors now both work for muffling gunshots (sound radius 1 instead of 3).

2017/5/22:
- Rewrote gunshot sound propagation to be breadth-first instead of depth-first. Behavior is now much more reliable.
- Modified mysql_wrapper to print the query to the console, but only if the DEBUG flag is enabled.

2017/5/21:
- Added $z parser flag, which shows the actor's desc, or prints their voice if you are unable to see them.
- Refactored MUD to describe magazines as magazines instead of clips.

src/olc.h

#ifndef _olc_h_
#define _olc_h_

#define NUM_MATERIALS           17
#define NUM_BARRIERS            9
#define NUM_AFF_FLAGS           22

// extern functions
extern void write_index_file(const char *suffix);

#define CH              d->character
#define CLS(ch)         send_to_char("\033[H\033[J", ch)
#define ZCMD            zone_table[zone].cmd[cmd_no]

#define VEDIT_CONFIRM_EDIT           0
#define VEDIT_MAIN_MENU              1
#define VEDIT_EDIT_NAMELIST          2
#define VEDIT_SHORTDESC              3
#define VEDIT_DESC                   4
#define VEDIT_LONGDESC               5
#define VEDIT_INSDESC       6
#define VEDIT_CONFIRM_SAVEDB         7
#define VEDIT_CONFIRM_SAVESTRING     8
#define VEDIT_PROMPT_APPLY           9
#define VEDIT_LEAVE	10
#define VEDIT_ARRIVE	11
#define VEDIT_ATT	12
#define VEDIT_HAND	13
#define VEDIT_SP	14
#define VEDIT_ACC	15
#define VEDIT_BOD	16
#define VEDIT_ARMOR	17
#define VEDIT_SIG	18
#define VEDIT_AUTO      19
#define VEDIT_LOAD      20
#define VEDIT_SEAT      21
#define VEDIT_SEAT2     22
#define VEDIT_COST      23
#define VEDIT_TYPE      24
#define VEDIT_PILOT	25
#define VEDIT_FLAGS	26
#define VEDIT_REARDESC	27
#define VEDIT_ENGINE		28

#define HEDIT_CONFIRM_EDIT	0
#define HEDIT_MAIN_MENU		1
#define HEDIT_TRIGGER		2
#define HEDIT_NAME		3
#define HEDIT_PARENT		4
#define HEDIT_KEYWORDS		5
#define HEDIT_DESC		6
#define HEDIT_SECURITY_COLOUR	7
#define HEDIT_SECURITY_RATING	8
#define HEDIT_SECURITY_DIFF	9
#define HEDIT_TYPE		10
#define HEDIT_CONFIRM_SAVESTRING 11
#define HEDIT_EXIT		12
#define HEDIT_RATINGS		13
#define HEDIT_RATINGS_ACCESS	14
#define HEDIT_RATINGS_CONTROL	15
#define HEDIT_RATINGS_INDEX	16
#define HEDIT_RATINGS_FILES	17
#define HEDIT_RATINGS_SLAVE	18
#define HEDIT_EXIT_ADD		19
#define HEDIT_EXIT_ADD2		20
#define HEDIT_EXIT_DEL		21
#define HEDIT_TRIGGER_DEL	22
#define HEDIT_TRIGGER_ADD	23
#define HEDIT_TRIGGER_ADD2	24
#define HEDIT_TRIGGER_ADD3	25
#define HEDIT_SSTOP		26
#define HEDIT_SSTART		27
#define HEDIT_EXTRA_MENU	28
#define HEDIT_EXTRA_ACCESS	29
#define HEDIT_EXTRA_CONTROL	30
#define HEDIT_EXTRA_INDEX	31
#define HEDIT_EXTRA_FILES	32
#define HEDIT_EXTRA_SLAVE	33
#define HEDIT_EXTRA_ACCESS2	34
#define HEDIT_EXTRA_FILES2	35
#define HEDIT_EXTRA_SLAVE2	36


#define ICEDIT_MAIN_MENU		0
#define ICEDIT_CONFIRM_EDIT		1
#define ICEDIT_CONFIRM_SAVESTRING 	2
#define ICEDIT_NAME			3
#define ICEDIT_ROOM			4
#define ICEDIT_DESC			5
#define ICEDIT_TYPE			6
#define ICEDIT_RATING			7
#define ICEDIT_SUBTYPE			8
#define ICEDIT_TRAP			9
#define ICEDIT_OPTION_MENU		10
#define ICEDIT_EXPERT			11

#define IEDIT_CONFIRM_EDIT           0
#define IEDIT_MAIN_MENU              1
#define IEDIT_EDIT_NAMELIST          2
#define IEDIT_SHORTDESC              3
#define IEDIT_DESC                   4
#define IEDIT_LONGDESC               5
#define IEDIT_TYPE                   6
#define IEDIT_EXTRAS                 7
#define IEDIT_WEAR                   8
#define IEDIT_WEIGHT                 9
#define IEDIT_COST                  10
#define IEDIT_TIMER                 12
#define IEDIT_VALUE_1               13
#define IEDIT_VALUE_2               14
#define IEDIT_VALUE_3               15
#define IEDIT_VALUE_4               16
#define IEDIT_VALUE_5               17
#define IEDIT_VALUE_6               18
#define IEDIT_VALUE_7               19
#define IEDIT_APPLY                 20
#define IEDIT_APPLYMOD              21
#define IEDIT_EXTRADESC_KEY         22
#define IEDIT_CONFIRM_SAVEDB        23
#define IEDIT_CONFIRM_SAVESTRING    24
#define IEDIT_PROMPT_APPLY          25
#define IEDIT_EXTRADESC_DESCRIPTION 26
#define IEDIT_EXTRADESC_MENU        27
#define IEDIT_MATERIAL                  28
#define IEDIT_RATING                    29
#define IEDIT_AFF_BITS                  30
#define IEDIT_VALUE_8                   31
#define IEDIT_VALUE_9                   32
#define IEDIT_VALUE_10                  33
#define IEDIT_AVAILTN			34
#define IEDIT_AVAILDAY			35
#define IEDIT_LEGAL1			36
#define IEDIT_LEGAL2			37
#define IEDIT_LEGAL3			38
#define IEDIT_VALUE_11                  39
#define IEDIT_VALUE_12                  40

/* Submodes of REDIT connectedness */
#define REDIT_CONFIRM_EDIT              0
#define REDIT_MAIN_MENU                 1
#define REDIT_NAME                      2
#define REDIT_DESC                      3
#define REDIT_FLAGS                     4
#define REDIT_SECTOR                    5
#define REDIT_EXIT_MENU                 6
#define REDIT_CONFIRM_SAVEDB            7
#define REDIT_CONFIRM_SAVESTRING        8
#define REDIT_EXIT_NUMBER               9
#define REDIT_EXIT_DESCRIPTION          10
#define REDIT_EXIT_KEYWORD              11
#define REDIT_EXIT_KEY                  12
#define REDIT_EXIT_KEY_LEV              13
#define REDIT_EXIT_DOORFLAGS            14
#define REDIT_EXTRADESC_MENU            15
#define REDIT_EXTRADESC_KEY             16
#define REDIT_EXTRADESC_DESCRIPTION     17
#define REDIT_EXIT_HIDDEN               19
#define REDIT_EXIT_BARRIER              20
#define REDIT_EXIT_MATERIAL             21
#define REDIT_MATRIX	                22
#define REDIT_LIBRARY_RATING		23
#define REDIT_HOST			29
#define REDIT_IO			30
#define REDIT_BASE			31
#define REDIT_TRACE			32
#define REDIT_ACCESS			33
#define REDIT_PARENT			34
#define REDIT_COMMLINK			35
#define REDIT_ADDRESS			36
#define REDIT_NDESC			37
#define REDIT_LIGHT			38
#define REDIT_SMOKE			39
#define REDIT_COMBAT			40
#define REDIT_COVER			41
#define REDIT_CROWD			42
#define REDIT_TYPE			43
#define REDIT_X				44
#define REDIT_Y				45
#define REDIT_Z				46
#define REDIT_BACKGROUND		47
#define REDIT_BACKGROUND2		48

#define MEDIT_CONFIRM_EDIT              0
#define MEDIT_CONFIRM_SAVESTRING        1
#define MEDIT_MAIN_MENU                 2
#define MEDIT_EDIT_NAMELIST             3
#define MEDIT_SHORT_DESCR               4
#define MEDIT_REG_DESCR                 5
#define MEDIT_LONG_DESCR                6
#define MEDIT_MOB_FLAGS                 7
#define MEDIT_AFF_FLAGS                 8
#define MEDIT_NUYEN                     10
#define MEDIT_EXPERIENCE                11
#define MEDIT_ATTRIBUTES                12
#define MEDIT_BOD                       13
#define MEDIT_QUI                       14
#define MEDIT_STR                       15
#define MEDIT_CHA                       16
#define MEDIT_INT                       17
#define MEDIT_WIL                       18
#define MEDIT_MAG                       19
#define MEDIT_BALLISTIC                 20
#define MEDIT_IMPACT                    21
#define MEDIT_PHYSICAL                  22
#define MEDIT_MENTAL                    23
#define MEDIT_POSITION                  24
#define MEDIT_DEFAULT_POSITION          25
#define MEDIT_GENDER                    26
#define MEDIT_WEIGHT                    27
#define MEDIT_HEIGHT                    28
#define MEDIT_LEVEL                     29
#define MEDIT_SKILLS                    30
#define MEDIT_SKILL1                    31
#define MEDIT_SKILL2                    32
#define MEDIT_SKILL3                    33
#define MEDIT_SKILL4                    34
#define MEDIT_SKILL5                    35
#define MEDIT_ATTACK_TYPE               36
#define MEDIT_SKILL1_VAL                37
#define MEDIT_SKILL2_VAL                38
#define MEDIT_SKILL3_VAL                39
#define MEDIT_SKILL4_VAL                40
#define MEDIT_SKILL5_VAL                41
#define MEDIT_ARRIVE_MSG                42
#define MEDIT_LEAVE_MSG                 43
#define MEDIT_CLASS                 44
#define MEDIT_CREDSTICK                 45

// quest defines
#define QEDIT_CONFIRM_EDIT                0
#define QEDIT_CONFIRM_SAVESTRING          1
#define QEDIT_MAIN_MENU                   2
#define QEDIT_JOHNSON                     3
#define QEDIT_TIME                        4
#define QEDIT_MIN_REP                     5
#define QEDIT_MAX_REP                     6
#define QEDIT_NUYEN                       7
#define QEDIT_KARMA                       8
#define QEDIT_M_MENU                      9
#define QEDIT_M_AWAIT_NUMBER             10
#define QEDIT_M_VNUM                     11
#define QEDIT_M_NUYEN                    12
#define QEDIT_M_KARMA                    13
#define QEDIT_M_LOAD                     14
#define QEDIT_M_OBJECTIVE                15
#define QEDIT_M_LDATA                    16
#define QEDIT_M_LDATA2                   17
#define QEDIT_M_ODATA                    18
#define QEDIT_O_MENU                     19
#define QEDIT_O_AWAIT_NUMBER             20
#define QEDIT_O_VNUM                     21
#define QEDIT_O_NUYEN                    22
#define QEDIT_O_KARMA                    23
#define QEDIT_O_LOAD                     24
#define QEDIT_O_OBJECTIVE                25
#define QEDIT_O_LDATA                    26
#define QEDIT_O_LDATA2                   27
#define QEDIT_O_ODATA                    28
#define QEDIT_INTRO                      29
#define QEDIT_DECLINE                    30
#define QEDIT_QUIT                       31
#define QEDIT_FINISH                     32
#define QEDIT_INFO                       33
#define QEDIT_REWARD                     34
#define QEDIT_SHOUR    35
#define QEDIT_EHOUR    36
#define QEDIT_SSTRING    37
#define QEDIT_ESTRING    38
#define QEDIT_DONE    39

// zone defines
#define ZEDIT_CONFIRM_EDIT_DATA         0
#define ZEDIT_CONFIRM_CREATE_DATA        1
#define ZEDIT_CONFIRM_ADD_CMD           2
#define ZEDIT_CONFIRM_EDIT_CMD          3
#define ZEDIT_DATA_MENU                 4
#define ZEDIT_ZONE_NAME                 5
#define ZEDIT_TOP_OF_ZONE               6
#define ZEDIT_LIFESPAN                  7
#define ZEDIT_RESET_MODE                8
#define ZEDIT_CONFIRM_SAVEDATA          9
#define ZEDIT_COMMAND_MENU              10
#define ZEDIT_CONFIRM_SAVECMDS          11
#define ZEDIT_IF_FLAG_CMD               12
#define ZEDIT_CMD_TYPE                  13
#define ZEDIT_ARG1                      14
#define ZEDIT_ARG2                      15
#define ZEDIT_ARG3                      16
#define ZEDIT_DIRECTION_OF_DOOR         17
#define ZEDIT_REMOVE_ROOM               18
#define ZEDIT_LOAD_ROOM                 19
#define ZEDIT_WEAR                      20
#define ZEDIT_DOOR_STATE                21
#define ZEDIT_CONFIRM_INSERT_CMD        22
#define ZEDIT_SECURITY                  23
#define ZEDIT_ID_LIST                   24
#define ZEDIT_CONNECTED                 25
#define ZEDIT_JURID			26

#define CEDIT_CONFIRM_EDIT        0
#define CEDIT_MAIN_MENU           1
#define CEDIT_CONFIRM_SAVESTRING  2
#define CEDIT_ALIAS               3
#define CEDIT_SHORT_DESC          4
#define CEDIT_DESC                5
#define CEDIT_LONG_DESC           6
#define CEDIT_VOICE    7
#define CEDIT_HEIGHT		  8
#define CEDIT_WEIGHT		  9
#define CEDIT_ARRIVE		10
#define CEDIT_LEAVE		11

#define SHEDIT_CONFIRM_EDIT	0
#define SHEDIT_MAIN_MENU	1
#define SHEDIT_BUYTYPES_MENU	2
#define SHEDIT_TEXT_MENU	3
#define SHEDIT_SELLING_MENU	4
#define SHEDIT_CONFIRM_SAVESTRING	5
#define SHEDIT_RACE_MENU	6
#define SHEDIT_FLAG_MENU	7
#define SHEDIT_KEEPER		8
#define SHEDIT_TYPE		9
#define SHEDIT_PROFIT_BUY	10
#define SHEDIT_PROFIT_SELL	11
#define SHEDIT_RANDOM		12
#define SHEDIT_OPEN		13
#define SHEDIT_OPEN2		14
#define SHEDIT_NSIK		15
#define SHEDIT_NSIP		16
#define SHEDIT_NEN		17
#define SHEDIT_NOBUY		18
#define SHEDIT_BUYMSG		19
#define SHEDIT_SELLMSG		20
#define SHEDIT_SELL_ADD		21
#define SHEDIT_SELL_ADD1	22
#define SHEDIT_SELL_ADD2	23
#define SHEDIT_SELL_DELETE	24
#define SHEDIT_ETTI		25
#define SHEDIT_SHOPNAME		26

#endif
